server/src/main/java/org/apache/lucene/index/BaseSingleStarTreeBuilder.java

```diff
@@ -0,0 +1,660 @@
+                String metricName = metricStatFieldPair.getField();
+                FieldInfo metricFieldInfos = state.fieldInfos.fieldInfo(metricName);
+                DocValuesType metricDocValuesType = state.fieldInfos.fieldInfo(metricName).getDocValuesType();
+                metricReaders[index] = starTreeDocValuesIteratorFactory.createIterator(
+                    metricDocValuesType,
+                    metricFieldInfos,
+                    docValuesProducer
+                );
+            }
+            index++;
+        }
+        this.maxLeafDocuments = starTreeFieldSpec.maxLeafDocs();
+    }
+
+    /**
+     * Generates the MetricStatFieldPairs for all the metrics on a field
+     *
+     * @return list of metric stat mapped with respective fields
+     */
+    public List<MetricStatFieldPair> generateMetricStatFieldPairs() {
+        List<MetricStatFieldPair> metricStatFieldPairs = new ArrayList<>();
+        for (Metric metric : this.starTreeField.getMetrics()) {
+            for (MetricStat metricType : metric.getMetrics()) {
+                MetricStatFieldPair metricStatFieldPair = new MetricStatFieldPair(metricType, metric.getField());
+                metricStatFieldPairs.add(metricStatFieldPair);
+            }
+        }
+        return metricStatFieldPairs;
+    }
+
+    /**
+     * Appends a star-tree document to the star-tree.
+     *
+     * @param starTreeDocument star tree document to be appended
+     */
+    public abstract void appendStarTreeDocument(StarTreeDocument starTreeDocument) throws IOException;
+
+    /**
+     * Returns the star-tree document of the given document id in the star-tree.
+     *
+     * @param docId Document dd
+     * @return Star tree document
+     */
+    public abstract StarTreeDocument getStarTreeDocument(int docId) throws IOException;
+
+    /**
+     * Returns the star-tree document of the given document id in the star-tree.
+     *
+     * @return Star tree document
+     */
+    public abstract List<StarTreeDocument> getStarTreeDocuments() throws IOException;
+
+    /**
+     * Returns the dimension value of the given document and dimension id in the star-tree.
+     *
+     * @param docId       Document Id
+     * @param dimensionId Dimension Id
+     * @return Dimension value
+     */
+    public abstract long getDimensionValue(int docId, int dimensionId) throws IOException;
+
+    /**
+     * Sorts and aggregates the star-tree Document in the segment, and returns a star-tree document iterator for all the
+     * aggregated star-tree document.
+     *
+     * <p>This method reads star-tree document from segment and generates the initial star-tree document for the star-tree.
+     *
+     * @param numDocs Number of documents in the segment
+     * @return Iterator for the aggregated star-tree document
+     */
+    public abstract Iterator<StarTreeDocument> processSegmentStarTreeDocuments(int numDocs) throws IOException;
```

Comment:
nit : processSegmentDocuments - this method does not process the star tree docs. It processes the normal documents of a segment and returns star tree docs iterator, lets fix the comments.

Reply:
ack!